src/Microsoft.Data.Analysis/DataFrame.IO.cs

```diff
@@ -510,5 +519,35 @@ public TextReader GetTextReader()
                 }
             }
         }
+        private static void WriteHeader(StreamWriter csvFile, IReadOnlyList<string> columnNames, char separator)
+        {
+            bool firstColumn = true;
+            var headerColumns = new StringBuilder();
```

Comment:
Why write to a StringBuilder, just to write the StringBuilder to the StreamWriter?

Why not just write directly to the StreamWriter as you go?

Reply:
I borrowed that design from the row-handling section of WriteCsv ([above](https://github.com/dotnet/machinelearning/pull/6303/files#diff-f44ef94ad825dc31ca6b7ab62c08f02f9d3eafab4b3b4f909b10f7e02e8ffcbfR458)), which uses StringBuilder in a similar way. Is there something about that case that makes StringBuilder appropriate that doesn't translate to this use of it in WriteHeader?

Reply:
Above, it reuses the `StringBuilder` over and over, so the cost of allocating the StringBuilder (and underlying `char[]` chunks) is amortized over many rows. Here we are only building up a single row - the header row.

It might be possible to also change the row-handling section above to not write to an intermediate StringBuilder and just write to the `StreamWriter` directly. However, looking quickly at it, it is using `AppendFormat(cultureInfo, "{0:...}", cell)`, which I don't think maps 1:1 with a StreamWriter API. StreamWriter doesn't appear to expose a CultureInfo in its formatting APIs.

Here we are just straight writing chars and strings, no need for formatting. It doesn't make sense to write it to an intermediate location, just to copy it again to its final spot.